x/ccv/provider/keeper/distribution.go

```diff
@@ -67,142 +66,265 @@ func (k Keeper) GetAllConsumerRewardDenoms(ctx sdk.Context) (consumerRewardDenom
 	return consumerRewardDenoms
 }
 
-// AllocateTokens performs rewards distribution to the community pool and validators
-// based on the Partial Set Security distribution specification.
-func (k Keeper) AllocateTokens(ctx sdk.Context) {
-	// return if there is no coins in the consumer rewards pool
-	if k.GetConsumerRewardsPool(ctx).IsZero() {
-		return
+// GetAllowlistedRewardDenoms returns the allowlisted reward denom for the given consumer id.
+func (k Keeper) GetAllowlistedRewardDenoms(ctx sdk.Context, consumerId string) ([]string, error) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.ConsumerIdToAllowlistedRewardDenomKey(consumerId))
+	if bz == nil {
+		return []string{}, nil
 	}
 
-	// Iterate over all launched consumer chains.
-	// To avoid large iterations over all the consumer IDs, iterate only over
-	// chains with an IBC client created.
-	for _, consumerId := range k.GetAllConsumersWithIBCClients(ctx) {
-		// note that it's possible that no rewards are collected even though the
-		// reward pool isn't empty. This can happen if the reward pool holds some tokens
-		// of non-whitelisted denominations.
-		alloc := k.GetConsumerRewardsAllocation(ctx, consumerId)
-		if alloc.Rewards.IsZero() {
-			continue
-		}
+	var denoms types.AllowlistedRewardDenoms
+	if err := denoms.Unmarshal(bz); err != nil {
+		return []string{}, err
+	}
+	return denoms.Denoms, nil
+}
 
-		chainId, err := k.GetConsumerChainId(ctx, consumerId)
+// SetAllowlistedRewardDenom sets the allowlisted reward denom for the given consumer id.
+func (k Keeper) SetAllowlistedRewardDenom(ctx sdk.Context, consumerId string, denom string) error {
+	currentDenoms, err := k.GetAllowlistedRewardDenoms(ctx, consumerId)
+	if err != nil {
+		return err
+	}
+	store := ctx.KVStore(k.storeKey)
+	updatedDenoms := types.AllowlistedRewardDenoms{Denoms: append(currentDenoms, denom)}
+	bz, err := updatedDenoms.Marshal()
+	if err != nil {
+		return err
+	}
+	store.Set(types.ConsumerIdToAllowlistedRewardDenomKey(consumerId), bz)
+	return nil
+}
+
+// DeleteAllowlistedRewardDenom deletes the allowlisted reward denom for the given consumer id.
+func (k Keeper) DeleteAllowlistedRewardDenom(ctx sdk.Context, consumerId string) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.ConsumerIdToAllowlistedRewardDenomKey(consumerId))
+}
+
+// UpdateAllowlistedRewardDenoms updates the allowlisted reward denoms for this consumer chain with the provided `rewardDenoms`
+func (k Keeper) UpdateAllowlistedRewardDenoms(ctx sdk.Context, consumerId string, rewardDenoms []string) error {
+	k.DeleteAllowlistedRewardDenom(ctx, consumerId)
+	for _, denom := range rewardDenoms {
```

Comment:
It seems that all of the denominations can be persisted at the same time since they're always deleted first.